torchdynamo/optimizations/distributed.py

```diff
@@ -0,0 +1,105 @@
+from typing import Dict
+from typing import List
+
+import torch
+import torch.fx as fx
+
+
+class DDPOptimizer:
+    def __init__(
+        self, bucket_bytes_cap: int, parameters_to_ignore: List[str], backend_compile_fn
+    ):
+        self.bucket_bytes_cap = bucket_bytes_cap
+        self.parameters_to_ignore = parameters_to_ignore
+        self.backend_compile_fn = backend_compile_fn
+
+    def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
```

Comment:
what happens with dynamic execution orders? Will re-compilation leads to new graph breaks? DDP might not be able to adapt to new bucketization accordingly, as 1) frequent rebucketziation is expensive 2) re-compilation is a one-time thing, but execution might switch between different orders multiple times.

Reply:
Let's assume that DDP initialized its buckets with the 'wrong' order, i.e. an order that will not be observed at runtime.

Also note, I don't think it's possible under dynamo for there to be dynamic behavior within one compiled function.  If something causes dynamism, it would force dynamo to recompile anyways.  This could be OK if it happened infrequently but otherwise it would be a perf issue (regardless of DDP) and the user might need to change their model to avoid it if using dynamo.

So, what I think would happen is
1) DDP's initial buckets are chosen. They may be wrong, and are currently ignored by DDPOptimizer.
2) DDPOptimizer partitions the graph into buckets according to the true order of execution _of the forward graph_.  
3) Then it runs AotAutograd and generates a backward graph, but bucket ordering is _not_ updated respecting the backward graph.
4) At runtime, the compiled fwd/bwd graph is run and hooks fire in-between the split subgraphs.  DDP may observe a different execution order and update its bucketing scheme after the first iteration.

If the program is not dynamic, things would be stable after that.  If there is dynamism, then I think each new version would incur dynamo compilation, and _if_ execution order is different enough to make DDP reconfigure buckets, it seems that DDP would do so multiple times.  

This seems to suggest that dynamo and DDP costs under dynamism are orthogonal: in eager, dynamism would potentially cause multiple re-configurations of DDP buckets, and that remains the same under dynamo.  Without DDP, dynamo still incurs expensive recompilation under dynamism.  Does that sound right to you @mrshenli?

Reply:
Yep, this makes sense to me. 

>  in eager, dynamism would potentially cause multiple re-configurations of DDP buckets, and that remains the same under dynamo.

Today's DDP will only reconfigure buckets once after the first iteration. But I agree with your point that, in this case, Dynamo + DDP should be able to offer on-par perf with eager-DDP. 

>  If there is dynamism, then I think each new version would incur dynamo compilation, and if execution order is different enough to make DDP reconfigure buckets, it seems that DDP would do so multiple times.

My original concern was that, suppose there are one dynamic if-else clause, and the app will switch between these two paths every 100 iterations. Would I be correct that dynamo will only trigger re-compile on the 1st and 101st iteration, but not 201st iteration as it is the same as the cached on for the 1st iteration? If that's the case, how can we tell DDP to update its buckets on the 201st iteration. Just OOC, I don't think this is a high-pri to address.

Reply:
> Would I be correct that dynamo will only trigger re-compile on the 1st and 101st iteration, but not 201st iteration as it is the same as the cached on for the 1st iteration?

yes, this is correct.  Dynamo would reuse cached programs for the if/else branches after they got compiled once each.

> how can we tell DDP to update its buckets on the 201st iteration.

Well, (1) dynamo's DDPOptimizer could insert a node into its compiled graph which calls into the DDP module and notifies it which bucket strategy it is about to use.  (2) this could happen on every iteration if we could make it fast enough and DDP mostly ignores it or checks its hash against the most recent one; or we could try to make dynamo actually pay attention to when its cached program changed from last cached program and only on those boundary conditions issue the notification to DDP.

I'm not sure how well this would work, but it seems like a possibility.  It might make more sense to invest in a new DDP approach that traces comm ops into the graph or implements DDP in FX though.  I see this graph-break solution as potentially a stop-gap.